src/index/hgraph_index.cpp

```diff
@@ -246,60 +248,20 @@ HGraphIndex::hnsw_add(const DatasetPtr& data) {
+                std::unique_lock<std::shared_mutex> wlock(this->global_mutex_);
                 for (int64_t j = max_level_; j <= level; ++j) {
                     this->route_graphs_.emplace_back(this->generate_one_route_graph());
                 }
                 max_level_ = level + 1;
-                need_lock = true;
-            } else {
-                lock.unlock();
-            }
-
-            {
-                auto ep = this->entry_point_id_;
-                MaxHeap result(allocator_);
-                for (auto j = max_level_ - 1; j > level; --j) {
-                    result = search_one_graph(
-                        datas + dim_ * i, route_graphs_[j], basic_flatten_codes_, ep, 1, nullptr);
-                    ep = result.top().second;
-                }
-
-                for (auto j = level; j >= 0; --j) {
-                    if (route_graphs_[j]->TotalCount() != 0) {
-                        result = search_one_graph(datas + dim_ * i,
-                                                  route_graphs_[j],
-                                                  basic_flatten_codes_,
-                                                  ep,
-                                                  this->ef_construct_,
-                                                  nullptr);
-                        ep = this->mutually_connect_new_element(
-                            inner_id, result, route_graphs_[j], basic_flatten_codes_, false);
-                    } else {
-                        route_graphs_[j]->InsertNeighborsById(inner_id,
-                                                              Vector<InnerIdType>(allocator_));
-                    }
-                    route_graphs_[j]->IncreaseTotalCount(1);
-                }
-                if (bottom_graph_->TotalCount() != 0) {
-                    result = search_one_graph(datas + dim_ * i,
-                                              this->bottom_graph_,
-                                              basic_flatten_codes_,
-                                              ep,
-                                              this->ef_construct_,
-                                              nullptr);
-                    this->mutually_connect_new_element(
-                        inner_id, result, this->bottom_graph_, basic_flatten_codes_, false);
-                } else {
-                    bottom_graph_->InsertNeighborsById(inner_id, Vector<InnerIdType>(allocator_));
-                }
-                bottom_graph_->IncreaseTotalCount(1);
-            }
-
-            if (need_lock) {
+                this->add_one_point(datas + i * dim_, level, inner_id);
                 entry_point_id_ = inner_id;
+                add_lock.unlock();
+            } else {
+                add_lock.unlock();
+                std::shared_lock<std::shared_mutex> rlock(this->global_mutex_);
+                this->add_one_point(datas + i * dim_, level, inner_id);
```

Comment:
add_one_point is thread safe ?

Reply:
no need global mutex, but have lock for per node's neighbors

Reply:
OK